File: Algorithmic Toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>
#define llu unsigned long long int
using namespace std;

llu get_fibonacci_huge_naive(llu n, llu m) {
    if (n <= 1)
        return n;

    llu previous = 0;
    llu current  = 1;

    for (llu i = 0; i < n - 1; ++i) {
        llu tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % m;
}


int pisano_period(int m)
{
	int prev = 0ull;
	int curr = 1ull;
	int tmp;
	int count=0;
	do
	{
		tmp  = curr;
		curr = curr + prev;
		prev = tmp;
		curr = curr%m;
		prev = prev%m;
		count++;
	}
	while(prev!=0 || curr!=1);
	return count;
}

int get_fibonacci_huge_fast(llu n,int m)
{
	int pp = pisano_period(m);
	int N = (int)(n%pp);

	if (N==0 || N==1)
		return N;

	int prev = 0;
	int curr = 1;
	while(N>1)
	{
		int temp = curr;
		curr = prev + curr;
		prev = temp;
		N--;
	}
	
	return curr%m;
}

int main() {
    llu n;
    int m;
    std::cin >> n >> m;
    //std::cout << get_fibonacci_huge_fast(n, m) << '\n';
    cout << pisano_period(m) << endl;
}

File: Algorithmic Toolbox/week2_algorithmic_warmup/5_fibonacci_number_again/fibonacci_huge.py
# Uses python3
import sys

def get_fibonacci_huge_naive(n, m):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m

def get_fibonacci_huge_fast(n, m):
	n = n % pisano_period(m)
	if n==0 or n==1 :
		return n
	prev = 0
	curr = 1
	while n>1 :
		tmp = curr;
		curr = prev+curr
		prev = tmp
		n -= 1
	return curr%m

def pisano_period(m) :
	prev = 0
	curr = 1
	count = 0
	while True :
		tmp = curr
		curr = prev + curr
		prev = tmp
		curr = curr%m
		prev = prev%m
		count += 1
		if prev==0 and curr==1 :
			break
	return count


if __name__ == '__main__':
    input = sys.stdin.read()
    n, m = map(int, input.split())
    #n,m = map(int,input().split())
    print(get_fibonacci_huge_fast(n, m))
    #print(pisano_period(m))
